### ISSUE TITLE
CURLDownloader downloads a zero byte file if download is unchanged (304)

### ISSUE THREAD

**Issue by hjuutilainen** [autopkg/autopkg#235]
I'm using OS X 10.11 and latest AutoPkg. This is easy to reproduce:

```
testvm:downloads testuser$ ls -la
total 16
drwxr-xr-x  3 testuser  staff   102 Nov 10 20:40 .
drwxr-xr-x  4 testuser  staff   136 Nov 10 20:36 ..
-rw-r--r--@ 1 testuser  staff  6148 Nov 10 20:41 .DS_Store
testvm:downloads testuser$ autopkg run -v TheUnarchiver.download
Processing TheUnarchiver.download...
SparkleUpdateInfoProvider
SparkleUpdateInfoProvider: Version retrieved from appcast: 103
SparkleUpdateInfoProvider: User-facing version retrieved from appcast: 3.10.1
SparkleUpdateInfoProvider: Found URL https://wakaba.c3.cx/releases/TheUnarchiver/TheUnarchiver3.10.1.zip
CURLDownloader
CURLDownloader: Storing new Last-Modified header: Wed, 23 Sep 2015 19:52:20 GMT
CURLDownloader: Storing new ETag header: "55fdeb-5206f73cfa500"
CURLDownloader: Downloaded /Users/testuser/Library/AutoPkg/Cache/com.github.autopkg.download.TheUnarchiver/downloads/TheUnarchiver3.10.1.zip
EndOfCheckPhase
Receipt written to /Users/testuser/Library/AutoPkg/Cache/com.github.autopkg.download.TheUnarchiver/receipts/TheUnarchiver-receipt-20151110-204115.plist

The following new items were downloaded:
    Download Path                                                                                                      
    -------------                                                                                                      
    /Users/testuser/Library/AutoPkg/Cache/com.github.autopkg.download.TheUnarchiver/downloads/TheUnarchiver3.10.1.zip  
testvm:downloads testuser$ ls -la
total 11024
drwxr-xr-x  4 testuser  staff      136 Nov 10 20:41 .
drwxr-xr-x  4 testuser  staff      136 Nov 10 20:36 ..
-rw-r--r--@ 1 testuser  staff     6148 Nov 10 20:41 .DS_Store
-rw-r--r--@ 1 testuser  staff  5635563 Nov 10 20:41 TheUnarchiver3.10.1.zip
testvm:downloads testuser$ autopkg run -v TheUnarchiver.download
Processing TheUnarchiver.download...
SparkleUpdateInfoProvider
SparkleUpdateInfoProvider: Version retrieved from appcast: 103
SparkleUpdateInfoProvider: User-facing version retrieved from appcast: 3.10.1
SparkleUpdateInfoProvider: Found URL https://wakaba.c3.cx/releases/TheUnarchiver/TheUnarchiver3.10.1.zip
CURLDownloader
CURLDownloader: Item at URL is unchanged.
CURLDownloader: Using existing /Users/testuser/Library/AutoPkg/Cache/com.github.autopkg.download.TheUnarchiver/downloads/TheUnarchiver3.10.1.zip
EndOfCheckPhase
Receipt written to /Users/testuser/Library/AutoPkg/Cache/com.github.autopkg.download.TheUnarchiver/receipts/TheUnarchiver-receipt-20151110-204120.plist

Nothing downloaded, packaged or imported.
testvm:downloads testuser$ ls -la
total 16
drwxr-xr-x  4 testuser  staff   136 Nov 10 20:41 .
drwxr-xr-x  4 testuser  staff   136 Nov 10 20:36 ..
-rw-r--r--@ 1 testuser  staff  6148 Nov 10 20:41 .DS_Store
-rw-r--r--@ 1 testuser  staff     0 Nov 10 20:41 TheUnarchiver3.10.1.zip
testvm:downloads testuser$ 
```

I don't know how this could be fixed other than doing two requests: only request the headers first and use an output file if actually downloading something.

**Comment by hjuutilainen** [comment]
Related curl issues:
https://github.com/bagder/curl/issues/237
https://github.com/bagder/curl/issues/183

**Comment by gregneagle** [comment]
Pull requests accepted. This was originally a proof-of-concept and yours is the first real feedback.

**Comment by gregneagle** [comment]
The other fix of course is to download to a temp path; if not unchanged delete the previous download and  rename, if unchanged delete the temp path

**Comment by hjuutilainen** [comment]
I'll take a stab at this. Using a temp file feels more hacky to me than checking the headers first (which also feels hacky). Do you think doing two different requests on new downloads would be acceptable?

**Comment by gregneagle** [comment]
I don't have a strong opinion. Usually, though, network (especially internet) operations are more "expensive" than local file system operations...

**Comment by timsutton** [comment]
Fixed by @hjuutilainen here: eb5130406871a02fbfa27c3ab2ef1c4a9811724b
